node/core/pvf/src/worker_common.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Common logic for implementation of worker processes.
+
+use crate::LOG_TARGET;
+use async_std::{
+	io,
+	os::unix::net::{UnixListener, UnixStream},
+	path::{PathBuf, Path},
+};
+use futures::{
+	AsyncRead, AsyncWrite, AsyncReadExt as _, AsyncWriteExt as _, FutureExt as _, never::Never,
+};
+use futures_timer::Delay;
+use rand::Rng;
+use std::{
+	fmt, mem,
+	pin::Pin,
+	task::{Context, Poll},
+	time::Duration,
+};
+use pin_project::pin_project;
+
+/// This is publicly exposed only for integration tests.
+#[doc(hidden)]
+pub async fn spawn_with_program_path(
+	debug_id: &'static str,
+	program_path: impl Into<PathBuf>,
+	extra_args: &'static [&'static str],
+	spawn_timeout: Duration,
+) -> Result<(IdleWorker, WorkerHandle), SpawnErr> {
+	let program_path = program_path.into();
+	with_transient_socket_path(debug_id, |socket_path| {
+		let socket_path = socket_path.to_owned();
+		async move {
+			let listener = UnixListener::bind(&socket_path)
+				.await
+				.map_err(|_| SpawnErr::Bind)?;
+
+			let handle = WorkerHandle::spawn(program_path, extra_args, socket_path)
+				.map_err(|_| SpawnErr::ProcessSpawn)?;
+
+			futures::select! {
+				accept_result = listener.accept().fuse() => {
+					let (stream, _) = accept_result.map_err(|_| SpawnErr::Accept)?;
+					Ok((IdleWorker { stream, pid: handle.id() }, handle))
+				}
+				_ = Delay::new(spawn_timeout).fuse() => {
+					Err(SpawnErr::AcceptTimeout)
+				}
+			}
+		}
+	})
+	.await
+}
+
+async fn with_transient_socket_path<R, F, Fut>(debug_id: &'static str, f: F) -> R
+where
+	F: FnOnce(&Path) -> Fut,
+	Fut: futures::Future<Output = R> + 'static,
+{
+	let socket_path = tmpfile(&format!("pvf-host-{}", debug_id));
+	let result = f(&socket_path).await;
+
+	// Best effort to remove the socket file. Under normal circumstances the socket will be removed
+	// by the worker. We make sure that it is removed here, just in case a failed rendezvous.
+	let _ = async_std::fs::remove_file(socket_path).await;
+
+	result
+}
+
+pub fn tmpfile(prefix: &str) -> PathBuf {
```

Comment:
What happens when the path already exists?

Why don't you just use the common crate for this? I think this maps down to some OS call?

Reply:
Yeah, initially I just didn't want to add yet another dependency, but since then the crate changed a lot. I will change it to tempfile.